replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,354 @@
+import com.connexta.replication.api.impl.NodeAdapters;
+import com.connexta.replication.api.impl.data.CreateRequestImpl;
+import com.connexta.replication.api.impl.data.CreateStorageRequestImpl;
+import com.connexta.replication.api.impl.data.ResourceRequestImpl;
+import com.connexta.replication.api.persistence.SiteManager;
+import com.connexta.replication.api.queue.Queue;
+import com.connexta.replication.data.MetadataImpl;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.net.URI;
+import java.sql.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles a single transfer between a local and remote Site based on the tasks received from the
+ * queue. This worker only supports harvesting operations. This worker will respond to thread
+ * interruptions.
+ */
+public class Worker extends Thread {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Worker.class);
+
+  private volatile boolean processing = false;
+
+  private volatile boolean running = true;
+
+  private final NodeAdapter local;
+
+  private final Queue queue;
+
+  private final SiteManager siteManager;
+
+  private final NodeAdapters nodeAdapters;
+
+  /**
+   * Creates a new worker which will monitor the provided queue and perform operations between the
+   * provided local and destination sites.
+   *
+   * <p>Does not begin monitoring of the queue until {@link #run()} is called.
+   *
+   * @param queue the queue to monitor tasks for
+   * @param local the local site's node adapter
+   * @param siteManager site manager for accessing sites
+   * @param nodeAdapters node adapter factory
+   */
+  Worker(Queue queue, NodeAdapter local, SiteManager siteManager, NodeAdapters nodeAdapters) {
+    this.queue = queue;
+    this.local = local;
+    this.siteManager = siteManager;
+    this.nodeAdapters = nodeAdapters;
+  }
+
+  /**
+   * Begin monitoring a queue for tasks. The handler of this worker is responsible for halting this
+   * thread by calling interrupt on this worker's thread, finishing the transfer of the current task
+   * if necessary.
+   */
+  @Override
+  public void run() {
+    try {
+      while (running) {
+        doRun();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } finally {
+      try {
+        local.close();
+      } catch (IOException ie) {
+        LOGGER.debug("Failed to close adapter", ie);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  void doRun() throws InterruptedException {
+    Task task = null;
+    try {
+      task = queue.take();
+      this.processing = true;
+      if (supported(task)) {
+        this.handle(task);
+      } else {
+        task.unlock();
+      }
+      this.processing = false;
+    } catch (InterruptedException e) {
+      if (task != null) {
+        try {
+          task.unlock();
```

Comment:
✏️ To avoid the nested try/catch, it might be worthwhile to add a new `unlockUnterruptibly()` method to the task which would do exactly that or again renamed the existing `unlock()` to `unlockInterruptibly()` and add a new `unlock()` that would do exactly that